doc/manual/src/contributing/hacking.md

```diff
@@ -8,50 +8,82 @@ $ git clone https://github.com/NixOS/nix.git
 $ cd nix
 ```
 
-To build Nix for the current operating system/architecture use
+## Nix with flakes
+
+This section assumes you are using Nix with flakes enabled. See the next section for equivalent instructions which don't require flakes.
+
+
+To build all dependencies and start a shell in which all environment
+variables are set up so that those dependencies can be found:
 
 ```console
-$ nix-build
+$ nix develop
 ```
 
-or if you have a flake-enabled nix:
+This shell also adds `./outputs/bin/nix` to your `$PATH` so you can run `nix` immediately after building it.
+
+To get a shell with a different compilation environment (e.g. `stdenv`,
+`gccStdenv`, `clangStdenv`, `clang11Stdenv`, `ccacheStdenv`):
```

Comment:
```suggestion
To get a shell with a different [compilation environment](#compilation-environments):
```

Let's list the environments and platforms in a separate section so we don't have to do double bookkeeping. I suggest putting that section after the two flakes/non-flakes sections.